### PR TITLE
Optimisations MeshSpray

### PR DESCRIPTION
**Pull request by ChuckAtShiro**

### CODE REVIEW

File: hide/comp/SceneEditor.hx [hunk 1]
```diff
@@ -2038,4 +2038,13 @@ class SceneEditor {
 		}
 		return null;
 	}
+
+	public function projectMeshSprays() {
+		var sprays : Array<hrt.prefab.l3d.MeshSpray> = [];
+		for(e in sceneData)
+			if(Std.is(e, hrt.prefab.l3d.MeshSpray))
+				sprays.push(cast e);
```
Comment (ncannasse) [inline hide/comp/SceneEditor.hx]: We need a more recursive lookup - sprays are not necessarily at toplevel. OTOH we don't want to lookup into MeshSpray so a custom recursive algorithm is necessary.

Comment (ChuckAtShiro) [reply hide/comp/SceneEditor.hx]: fixed

**Comment by ncannasse** [comment]
@ChuckAtShiro the PR now conflicts ;)

File: hide/comp/SceneEditor.hx [hunk 2]
```diff
@@ -365,6 +365,7 @@ class SceneEditor {
 			return true;
 		}
 		tree.get = function(o:PrefabElement) {
+			if(Std.is(o, hrt.prefab.l3d.MeshSpray)) return [];
```
Comment (trethaller) [inline hide/comp/SceneEditor.hx]: @ncannasse wouldn't we want a more generic way to do that ? Something like `Prefab.getDisplayChildren()` or something? Not too fond of having this here

File: hide/comp/SceneEditor.hx [hunk 3]
```diff
@@ -758,21 +760,24 @@ class SceneEditor {
...
 						refreshProps();
 					}
-
 					for(o in objects3d)
 						o.updateInstance(getContext(o));
 				}));
-
 				for(o in objects3d)
 					o.updateInstance(getContext(o));
+        for(i in 0...objects3d.length) {
```
Comment (trethaller) [inline hide/comp/SceneEditor.hx]: Why not putting this in the `for(o in objects3d)` above ?

File: hide/comp/SceneEditor.hx [hunk 4]
```diff
@@ -758,21 +760,24 @@ class SceneEditor {
...
+					var sprayObj = Std.downcast(sceneObjs[i].parent, hrt.prefab.l3d.SprayObject);
+					if(sprayObj != null) {
+						@:privateAccess sprayObj.blockHead = null;
```
Comment (trethaller) [inline hide/comp/SceneEditor.hx]: Why not doing this directly in `SprayObject.updateInstance`?

File: hide/comp/SceneEditor.hx [hunk 5]
```diff
@@ -1070,18 +1077,26 @@ class SceneEditor {
...
 
 		var map = new Map<PrefabElement,Bool>();
-		function selectRec(e : PrefabElement, b:Bool) {
+		function select(e : PrefabElement, b:Bool) {
 			if( map.exists(e) )
 				return;
 			map.set(e, true);
 			var ectx = context.shared.contexts.get(e);
 			setObjectSelected(e, ectx == null ? context : ectx, b);
+		}
```
Comment (trethaller) [inline hide/comp/SceneEditor.hx]: @ncannasse what do you think of making `setObjectSelected` return false if we want to stop the recursive selection?

File: hrt/prefab/Prefab.hx [hunk 6]
```diff
@@ -311,14 +311,22 @@ class Prefab {
...
+		Find a single prefab in the tree by calling `f` on each and returning the first not-null value returned, or null if not found. Search is breadth first.
 	**/
 	public function find<T>( f : Prefab -> Null<T> ) : Null<T> {
 		var v = f(this);
 		if( v != null )
 			return v;
+		return this.doFind(f);
+	}
+
+	public function doFind<T>( f : Prefab -> Null<T> ) : Null<T> {
```
Comment (trethaller) [inline hrt/prefab/Prefab.hx]: Should this be public?

File: hide/comp/SceneEditor.hx [hunk 7]
```diff
@@ -562,6 +561,9 @@ class SceneEditor {
 			if(c != null) c;
 		}]);
 		var boundsCollider = new h3d.col.ObjectCollider(local3d, bounds);
+		var r = Math.max(bounds.getSize().z, Math.max(bounds.getSize().x, bounds.getSize().y));
+		var pos = local3d.getAbsPos();
+		local3d.cullingCollider = new h3d.col.Sphere(pos.tx, pos.ty, pos.tz, r);
```
[the remark this answers is not in the document]
Comment (ChuckAtShiro) [reply hide/comp/SceneEditor.hx]: Let me fix this for you.

File: hide/comp/SceneEditor.hx (hunk 3 above)
Comment (ChuckAtShiro) [reply hide/comp/SceneEditor.hx]: Let me fix this for you,

File: hrt/prefab/Prefab.hx (hunk 6 above)
Comment (ChuckAtShiro) [reply hrt/prefab/Prefab.hx]: Let me fix this for you,

**Comment by ChuckAtShiro** [comment]
Bigger fixes coming soon.

File: hide/comp/SceneEditor.hx [hunk 8]
```diff
@@ -563,7 +563,7 @@ class SceneEditor {
 		var boundsCollider = new h3d.col.ObjectCollider(local3d, bounds);
 		var r = Math.max(bounds.getSize().z, Math.max(bounds.getSize().x, bounds.getSize().y));
 		var pos = local3d.getAbsPos();
-		local3d.cullingCollider = new h3d.col.Sphere(pos.tx, pos.ty, pos.tz, r);
+		local3d.cullingCollider = new h3d.col.ObjectCollider(local3d, new h3d.col.Sphere(0, 0, 0, r));
```
Comment (trethaller) [inline hide/comp/SceneEditor.hx]: Cool, then `pos` is unused right?

Comment (ChuckAtShiro) [reply hide/comp/SceneEditor.hx]: Let me fix this for you.

**Pull request closed without merging**